dateutil/tz/tz.py

```diff
@@ -1459,7 +1459,8 @@ def __get_gettz():
 
     class GettzFunc(object):
         """
-        gettz takes a variety of inputs and returns a time zone.
+        Given a variety of inputs, like time zone name and ``TZ`` varible,
+        return a time zone.
 
         :param object:
```

Comment:
This is not actually the interface. `GettzFunc` is a class representing a callable function. `object` is the name of the base type of all objects in Python 3 ("new style classes"), and so `GettzFunc` is a subclass of `object` for backwards compatibility with Python 2.

I realize it's confusing, but what's being written here is *actually* the documentation for `tz.gettz`, which is an instance of `GettzFunc`, but for practical purposes is equivalent to a `function`.

When you call `tz.gettz()`, it actually invokes [`GettzFunc.__call__`](https://github.com/weatherpattern/dateutil/blob/503b5e9331a4d800bb5875d02612b097c57d12de/dateutil/tz/tz.py#L1501), which itself is a thin wrapper around [`GettzFunc.nocache`](https://github.com/weatherpattern/dateutil/blob/503b5e9331a4d800bb5875d02612b097c57d12de/dateutil/tz/tz.py#L1520).

The reason it's built this way is partially for performance reasons, but mostly because I want `tz.gettz` to satisfy the following invariant:

```python
assert tz.gettz(name) is tz.gettz(name)
```

So I need to store a reference to the arguments that have been passed to `tz.gettz` and all the results, so that I can return the same object if I get the same arguments. Those get stored on the `GettzFunc` instance in the `_cache`. Because you might want the old semantics, I've exposed a "nocache" [variant](https://github.com/python-variants/variants) that bypasses the cache: `tz.gettz.nocache(name)`, and since cache misses are identical to calls to `nocache`, `__call__` is implemented in terms of `nocache`.

There are also functions on `gettz` for manipulating the cache, like `clear_cache`.

Reply:
Thanks for the extended comments. I'm going to a little more time digest and a bit more research and make another pass. More coming.

Reply:
OK I took another pass and hopefully more accurate now. I also added the assertion as an example.

A couple questions:
Your explanation above was really helpful. Do you want me to make a new issue to put that information into the individual methods and functions as comments?

Also, I wanted to double check:
```
>>tz.gettz() is tz.gettz()
False
```
This is as designed, right? Because if ```name=None```, then the time zone is not cached.

Reply:
> Do you have an example (could be anywhere) of how much detail you prefer in the docstring?

I'm not really satisfied with the `dateutil` documentation in general. I personally originated the functions (and hence the docstrings) for [isoparse](https://dateutil.readthedocs.io/en/stable/parser.html#dateutil.parser.isoparse) and [default_tzinfo](https://dateutil.readthedocs.io/en/stable/utils.html#dateutil.utils.default_tzinfo), so something like that is likely what I would write, but I actually need to do a full re-organization of the documentation, and I'm not sure where information should go for each function / class.

I'm thinking in the end that however it's organized, *somewhere* in the documentation, the functions and classes would be documented roughly with a `man` page level of detail.

> Your explanation above was really helpful. Do you want me to make a new issue to put that information into the individual methods and functions as comments?

Sure.

> Also, I wanted to double check:

```python
>>> tz.gettz() is tz.gettz
False
```
> This is as designed (I think.) Because if name=None, then the time zone is not cached. Is that correct?

Yes and no. Ideally that would be `True` even for `None`, but since `tz.gettz()` represents local time it's made a bit more complicated by the fact that local time might change over the runtime of a program (which leads to a bunch of bugs, but we don't want to add to the bugginess by adding to the problem). As such, I'd rather err on the side of the previous non-cached behavior until we can work out how best to figure out if the "local zone" cache has been invalidated by changes in the execution environment.